### ISSUE TITLE
Let RequestMaxSize have effect on all requests?

### ISSUE THREAD

**Issue by jonasps** [hossein1376/grape#6]
Hello, thank you for your work.

I was thinking, would it make sense to let the Option `RequestMaxSize` have effect on all request and not just affecting the default Serializer.

Let's say it would be part of the server
```
type Server struct {
	Logger
	Serializer
	response
	RequestMaxSize int64
}
```

Then It could  be used for all requests.

Here is an example of how it could be used by a middleware (which I'm not sure is the best idea but anyway)

```
// MaxSizeMiddleware will abort requests if request body size exceeds server.RequestMaxSize.
func (server Server) MaxSizeMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var size int64
		b, _ := ioutil.ReadAll(r.Body)
		for range b {
			size += 1
		}
		if size > server.RequestMaxSize {
			server.Response(w, 413, "Content Too Large")
		} else {
			next.ServeHTTP(w, r)
		}
	})
}

```
What's your opinion about this?

**Comment by hossein1376** [comment]
Hey, thanks for your suggestion!

Standard library suggests using [MaxBytesReader](https://pkg.go.dev/net/http#MaxBytesReader).
I think reading _every_ request's body, like the middleware you provided or some other implementation, will have a (possibly huge) performance overhead for little gain. I believe we should enforce this limitation only when we actually care about the body, i.e., reading the body for json inputs. 

Adding an O(N) operation just to be on the safer side doesn't seem like a desirable option. Let me know if I'm wrong about the benefits of your approach or if there's a security risk involved :)

**Comment by jonasps** [comment]
Ok, **MaxBytesReader** is likely the right thing to use in this case.  The Middleware snippet probably made my question more confusing so ignore that.

Making MaxBytesReader a respected setting no matter what serializer I use is actually the main question, and one way to achieve this would be to    let the server check the request size instead of the serializer.

**Comment by hossein1376** [comment]
Decoupling RequestMaxSize and serializer is a really good idea. Achieving that without affecting performance too much is the problem here.  

Including MaxBytesReader in a middleware may not be that big of a deal. Maybe doing a load test can helpful? If you're interested, [k6](https://github.com/grafana/k6) is a fantastic tool. I'll run my own tests as soon as I get the chance.

**Comment by hossein1376** [comment]
I finally did the load test! The setup was for 10 virtual users to send as many post requests (with 1 megabyte body size) as they can in 30 seconds.  The following are the numbers on my machine.

While reading the body, Grape did 255.05 req/s and 7.7k total. In turns out, `MaxBytesReader` is really performant and creates next to zero overhead, as removing it barely made an improvement.

While ignoring the body in the handler, the current implementation did 1.84k req/s and 55.3k overall. Then, I moved the `MaxBytesReader` call to a middleware and the result was 1.62k req/s and 48.8k total. That's a ~**10%** performance decrease.  I repeated the test multiple times and the results were the similar.

Granted, this is an extreme case and the impact will be much lower on common use cases. Still, I don't think applying body size limitation on *all* requests with a middleware is a good idea. 

My real concern is, if we proceed with this approach, there's the possibility for users to forget to include the middleware. This sort of thing happens all the time and Grape as a library should help users avoid vulnerabilities, not causing them.  
A possible solution is to put a flag inside context in the middleware. Then, check for it in the `ReadJson` function and if not present, apply size limit. Although, this introduces unneeded complexity and I'm reluctant to do so.

I appreciate your suggestion, but I just don't think there is any good way to implement this. RequestMaxSize is tightly coupled with the `ReadJson` function and separating them just doesn't worth the trouble.
